src/web/extractors.rs

```diff
@@ -0,0 +1,107 @@
+//! Request header/body/query extractors
+//!
+//! Handles ensuring the header's, body, and query parameters are correct, extraction to
+//! relevant types, and failing correctly with the appropriate errors if issues arise.
+use actix_web::{Error, FromRequest, HttpRequest, Path, State};
+use futures::{future, Future};
+use serde::de::{Deserialize, Deserializer};
+
+use db::{params, Db};
+use server::ServerState;
+use settings::Settings;
+use web::auth::{HawkIdentifier, HawkPayload};
+
+// XXX: Convert these to full extractors, or consider validator functions if extractors isn't going
+//      to work.
+pub type MetaRequest = (Path<UidParam>, HawkIdentifier, State<ServerState>);
+pub type GetCollectionRequest = (Path<CollectionParams>, HawkIdentifier, State<ServerState>);
+pub type BsoRequest = (Path<BsoParams>, HawkIdentifier, State<ServerState>);
+
+/// Request arguments needed for Information Requests
+///
+/// Only the database and user identifier is required for information
+/// requests: https://mozilla-services.readthedocs.io/en/latest/storage/apis-1.5.html#general-info
+/*
+pub struct MetaRequest<'a> {
+    pub db: Box<&'a Db>,
+    pub user_id: HawkIdentifier,
+}
+
+impl <'a>FromRequest<ServerState> for MetaRequest<'a> {
+    type Config = Settings;
+    type Result = Box<Future<Item = MetaRequest<'a>, Error = Error>>;
+
+    fn from_request(req: &HttpRequest<ServerState>, cfg: &Self::Config) -> Self::Result {
+        Box::new(
+            <(Path<UidParam>, HawkPayload, State<ServerState>)>::extract(req)
+                .and_then(|(path, auth, state)| {
+                    future::ok(MetaRequest {
+                        db: Box::new(db),
+                        user_id: auth.user_id,
+                    })
+                }),
+        )
+    }
+}
+*/
+
+#[derive(Deserialize)]
+pub struct UidParam {
+    uid: String,
+}
+
+#[derive(Deserialize)]
+pub struct DeleteCollectionQuery {
+    pub ids: Option<BsoIds>,
+}
+
+pub struct BsoIds(pub Vec<String>);
+
+impl<'d> Deserialize<'d> for BsoIds {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'d>,
+    {
+        let value: String = Deserialize::deserialize(deserializer)?;
+        Ok(BsoIds(value.split(",").map(|id| id.to_string()).collect()))
+    }
+}
+
+#[derive(Deserialize, Serialize)]
+pub struct PostCollectionBody {
+    pub id: String,
+    pub sortindex: Option<i32>,
+    pub payload: Option<String>,
+    pub ttl: Option<u32>,
+}
+
+impl From<PostCollectionBody> for params::PostCollectionBso {
```

Comment:
will this extractor include more validation later?

The PostCollectionBody struct is a duplicate of PostCollectionBso (minus the ability to Deserialize) -- wondering if we can get rid of the duplication

Reply:
No, the validation should really occur in the extractor, this From looks like its just a transform. I'm not entirely sure why its needed at all.